src/dynamicprompts/samplers/command_collection.py

```diff
@@ -0,0 +1,46 @@
+from __future__ import annotations
+
+from typing import Iterable
+
+from dynamicprompts.commands import Command
+from dynamicprompts.samplers.base import SamplerRouter
+from dynamicprompts.types import CommandList, StringGen
+
+
+class CommandCollection:
+    """
+    A class that holds a collection of commands and manages generating values for them.
+    """
+
+    def __init__(self, commands: Iterable[Command], sampler_router: SamplerRouter):
+        if sampler_router is None:
+            raise ValueError("Sampler router must be provided")
+
+        self._commands = list(commands)
+        self._generators = [
+            sampler_router.generator_from_command(c) for c in self._commands
+        ]
+        self._values: list[str | None] = [next(g) for g in self._generators]
+        self._sampler_router = sampler_router
+
+    def get_value(self, command: Command) -> str | None:
+        if command not in self._commands:
+            raise ValueError(f"Command {command} not in collection")
+        index = self._commands.index(command)
```

Comment:
`.index` raises a `ValueError` when the value isn't found anyway, so this now does an additional O(n) list traversal.